src/main/java/plugily/projects/buildbattle/commands/TabCompletion.java

```diff
@@ -1,66 +1,107 @@
- * <p>
- * Created at 27.05.2018
- */
 public class TabCompletion implements TabCompleter {
+    private final HashMultimap<String, String> completesListMap;
 
-  private ArgumentsRegistry registry;
-
-  public TabCompletion(ArgumentsRegistry registry) {
-    this.registry = registry;
-  }
+    public TabCompletion() {
+        completesListMap = HashMultimap.create();
+        add("bb", "join", "randomjoin", "stats", "leave", "top", "create");
+        add("bba", "list", "stop", "forcestart", "reload", "delete", "addplot", "removeplot", "addnpc", "settheme", "votes", "plotwand");
+        add("arena", "[arena]");
+        add("soloteam", "[solo/team]");
+        add("onlineplayer", "(online player)");
+        add("statistic", "[statistic]");
+        add("theme", "(theme)");
+        add("plotID", "[plot ID]");
+        add("addset", "[add/set]");
+        add("amount", "[amount]");
+        add("player", "(player)");
+    }
 
-  @Override
-  public List<String> onTabComplete(CommandSender sender, Command cmd, String label, String[] args) {
-    if (!(sender instanceof Player)) {
-      return Collections.emptyList();
+    private void add(String key, String... args) {
+        completesListMap.putAll(key, Arrays.asList(args));
     }
-    if (cmd.getName().equalsIgnoreCase("buildbattleadmin") && args.length == 1) {
-      return registry.getMappedArguments().get(cmd.getName().toLowerCase()).stream().map(CommandArgument::getArgumentName).collect(Collectors.toList());
+
+    private List<String> getPartial(String token, Iterable<String> collection) {
+        return StringUtil.copyPartialMatches(token, collection, new ArrayList<>());
     }
-    if (cmd.getName().equalsIgnoreCase("buildbattle")) {
-      if (args.length == 2 && args[0].equalsIgnoreCase("join")) {
-        return ArenaRegistry.getArenas().stream().map(BaseArena::getID).collect(Collectors.toList());
-      }
-      if (args.length == 1) {
-        return registry.getMappedArguments().get(cmd.getName().toLowerCase()).stream().map(CommandArgument::getArgumentName).collect(Collectors.toList());
-      }
+
+    public
+    @Override List<String> onTabComplete(@NotNull CommandSender sender, @NotNull Command cmd, @NotNull String label, String[] args) {
+        if (!(sender instanceof Player)) {
+            return Collections.emptyList();
+        }
+        switch (args.length) {
+            case 1:
+                if (label.equalsIgnoreCase("bb") || label.equalsIgnoreCase("buildbattle")) {
+                    return getPartial(args[0], completesListMap.get("bb"));
+                }
+                if (label.equalsIgnoreCase("bba") || label.equalsIgnoreCase("buildbattleadmin")) {
+                    return getPartial(args[0], completesListMap.get("bba"));
+                }
+            case 2:
+                if (completesListMap.get("bb").contains(args[0].toLowerCase())) {
+                    if (args[0].equalsIgnoreCase("join") || args[0].equalsIgnoreCase("create")) {
```

Comment:
Would be good to use `switch (args[0].toLowerCase())` for better code

Reply:
I know that my code is very messy, and kinda bad, but the tab completers are better

Reply:
? wat
What I'm suggested is just a code-cleanup and makes more readable.

Reply:
I know, I'm just saying that I'm very bad at cleaning up code

Reply:
I'm going to try to make it so it doesn't affect the other files... it might take some time tho

Reply:
I just uploaded the class to hastebin, and I just redownloaded the source files